### ISSUE TITLE
Commit 9c91314886 breaks YVES/Sereal-Decoder-4.018.tar.gz on threaded perl

### ISSUE THREAD

**Issue by andk** [Perl/perl5#19382]
**Description**

During compilation of YVES/Sereal-Decoder-4.018.tar.gz with a threaded perl built with v5.35.5-32-g9c91314886 the following error is thrown:
``` text
cc -c  -I. -D_REENTRANT -D_GNU_SOURCE -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2 -O3 -Werror=declaration-after-statement    -DVERSION=\"4.018\" -DXS_VERSION=\"4.018\" -fPIC "-I/home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.35.5-32-g9c91314886/1e0c/lib/5.35.6/x86_64-linux-thread-multi/CORE"  -DNDEBUG -DHAVE_ZSTD Decoder.c

In file included from /home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.35.5-32-g9c91314886/1e0c/lib/5.35.6/x86_64-linux-thread-multi/CORE/perl.h:7135,
                 from Decoder.xs:6:
/home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.35.5-32-g9c91314886/1e0c/lib/5.35.6/x86_64-linux-thread-multi/CORE/inline.h: In function ‘Perl_cop_file_avn’:
/home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.35.5-32-g9c91314886/1e0c/lib/5.35.6/x86_64-linux-thread-multi/CORE/inline.h:3439:5: error: ISO C90 forbids mixed declarations and code [-Werror=declaration-after-statement]
 3439 |     const char *file = CopFILE(cop);
      |     ^~~~~
cc1: some warnings being treated as errors
make: *** [Makefile:368: Decoder.o] Error 1

```

I see same result with v5.35.8-28-gebfc9d5d0d given a threaded perl.

**Steps to Reproduce**

cpan -i Sereal::Decoder

**Expected behavior**

Module should be built, tested and installed

**Perl configuration**

```
Summary of my perl5 (revision 5 version 35 subversion 6) configuration:
  Commit id: 9c913148860b0e83e6149d37e86cdb29663ee812
  Platform:
    osname=linux
    osvers=5.15.0-2-amd64
    archname=x86_64-linux-thread-multi
    uname='linux k93msid 5.15.0-2-amd64 #1 smp debian 5.15.5-2 (2021-12-18) x86_64 gnulinux '
    config_args='-Dprefix=/home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.35.5-32-g9c91314886/1e0c -Dmyhostname=k93msid -Dinstallusrbinperl=n -Uversiononly -Dusedevel -des -Ui_db -Dlibswanted=cl pthread socket inet nsl gdbm dbm malloc dl ld sun m crypt sec util c cposix posix ucb BSD gdbm_compat -Duseithreads -Uuselongdouble -DEBUGGING=-g'
    hint=recommended
    useposix=true
    d_sigaction=define
    useithreads=define
    usemultiplicity=define
    use64bitint=define
    use64bitall=define
    uselongdouble=undef
    usemymalloc=n
    default_inc_excludes_dot=define
  Compiler:
    cc='cc'
    ccflags ='-D_REENTRANT -D_GNU_SOURCE -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2'
    optimize='-O2 -g'
    cppflags='-D_REENTRANT -D_GNU_SOURCE -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include'
    ccversion=''
    gccversion='11.2.0'
    gccosandvers=''
    intsize=4
    longsize=8
    ptrsize=8
    doublesize=8
    byteorder=12345678
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=16
    longdblkind=3
    ivtype='long'
    ivsize=8
    nvtype='double'
    nvsize=8
    Off_t='off_t'
    lseeksize=8
    alignbytes=8
    prototype=define
  Linker and Libraries:
    ld='cc'
    ldflags =' -fstack-protector-strong -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib/x86_64-linux-gnu /usr/lib /lib/x86_64-linux-gnu /lib
    libs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    perllibs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    libc=libc-2.33.so
    so=so
    useshrplib=false
    libperl=libperl.a
    gnulibc_version='2.33'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs
    dlext=so
    d_dlsymun=undef
    ccdlflags='-Wl,-E'
    cccdlflags='-fPIC'
    lddlflags='-shared -O2 -g -L/usr/local/lib -fstack-protector-strong'


Characteristics of this binary (from libperl): 
  Compile-time options:
    HAS_TIMES
    MULTIPLICITY
    PERLIO_LAYERS
    PERL_COPY_ON_WRITE
    PERL_DONT_CREATE_GVSV
    PERL_MALLOC_WRAP
    PERL_OP_PARENT
    PERL_PRESERVE_IVUV
    PERL_USE_DEVEL
    USE_64_BIT_ALL
    USE_64_BIT_INT
    USE_ITHREADS
    USE_LARGE_FILES
    USE_LOCALE
    USE_LOCALE_COLLATE
    USE_LOCALE_CTYPE
    USE_LOCALE_NUMERIC
    USE_LOCALE_TIME
    USE_PERLIO
    USE_PERL_ATOF
    USE_REENTRANT_API
    USE_THREAD_SAFE_LOCALE
  Built under linux
  Compiled at Jan 30 2022 07:15:24
  @INC:
    /home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.35.5-32-g9c91314886/1e0c/lib/site_perl/5.35.6/x86_64-linux-thread-multi
    /home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.35.5-32-g9c91314886/1e0c/lib/site_perl/5.35.6
    /home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.35.5-32-g9c91314886/1e0c/lib/5.35.6/x86_64-linux-thread-multi
    /home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.35.5-32-g9c91314886/1e0c/lib/5.35.6

```

**Comment by jkeenan** [comment]
> During compilation of YVES/Sereal-Decoder-4.018.tar.gz with a threaded perl built with v5.35.5-32-g9c91314886 the following error is thrown:
> 
> ```
> cc -c  -I. -D_REENTRANT -D_GNU_SOURCE -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2 -O3 -Werror=declaration-after-statement    -DVERSION=\"4.018\" -DXS_VERSION=\"4.018\" -fPIC "-I/home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.35.5-32-g9c91314886/1e0c/lib/5.35.6/x86_64-linux-thread-multi/CORE"  -DNDEBUG -DHAVE_ZSTD Decoder.c
> 
> In file included from /home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.35.5-32-g9c91314886/1e0c/lib/5.35.6/x86_64-linux-thread-multi/CORE/perl.h:7135,
>                  from Decoder.xs:6:
> /home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.35.5-32-g9c91314886/1e0c/lib/5.35.6/x86_64-linux-thread-multi/CORE/inline.h: In function ‘Perl_cop_file_avn’:
> /home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.35.5-32-g9c91314886/1e0c/lib/5.35.6/x86_64-linux-thread-multi/CORE/inline.h:3439:5: error: ISO C90 forbids mixed declarations and code [-Werror=declaration-after-statement]
>  3439 |     const char *file = CopFILE(cop);
>       |     ^~~~~
> cc1: some warnings being treated as errors
> make: *** [Makefile:368: Decoder.o] Error 1
> ```
> 
> I see same result with v5.35.8-28-gebfc9d5d0d given a threaded perl.

This problem is at least in part dependent on the C-compiler used to build the `perl` against which `Sereal::Decoder` is being installed.  I've been able to reproduce the problem in `perl`s built with `gcc`, but with `perl`s built with `clang` I have no problem installing `Sereal::Decoder`.

On FreeBSD-12, I built a threaded `perl` as I customarily do, _i.e.,_ using the default C-compiler `clang` (version 10.0.1).
```
$ ./bin/perl -v | head -2 | tail -1
This is perl 5, version 35, subversion 9 (v5.35.9 (v5.35.8-29-g4ccd1515c8)) built for amd64-freebsd-thread-multi
$ ./bin/perl -Ilib -V:config_args
config_args='-des -Dusedevel -Uversiononly -Dprefix=/home/jkeenan/testing/blead -Dman1dir=none -Dman3dir=none -Duseithreads -Doptimize=-O2 -pipe -fstack-protector -fno-strict-aliasing';
$ ./bin/perl -Ilib -V:gccversion
gccversion='FreeBSD Clang 10.0.1 (git@github.com:llvm/llvm-project.git llvmorg-10.0.1-0-gef32c611aa2)';
# Then after using 'cpan' to install Sereal::Decoder ...
$ ./bin/perl -Ilib -MSereal::Decoder -E 'say qq|$Sereal::Decoder::VERSION|;'
4.018
```
So, no problem installing `Sereal::Decoder` against threaded blead perl built with `clang`.  Now let's try `gcc` version 12, threaded build at the very same commit.
```
$ ./bin/perl -v | head -2 | tail -1
This is perl 5, version 35, subversion 9 (v5.35.9 (v5.35.8-29-g4ccd1515c8)) built for amd64-freebsd-thread-multi
$ ./bin/perl -Ilib -V:config_args
config_args='-des -Dusedevel -Uversiononly -Dprefix=/home/jkeenan/testing/blead -Dman1dir=none -Dman3dir=none -Duseithreads -Doptimize=-O2 -pipe -fstack-protector -fno-strict-aliasing -Dcc=gcc12';
$ ./bin/perl -Ilib -V:gccversion
gccversion='12.0.0 20210919 (experimental)';
$ ./bin/cpan -i Sereal::Decoder
...
gcc12 -c  -I.  -DHAS_FPSETMASK -DHAS_FLOATINGPOINT_H -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_FORTIFY_SOURCE=2 -O3 -Werror=declaration-after-statement    -DVERSION=\"4.018\"  -DXS_VERSION=\"4.018\" -DPIC -fPIC "-I/home/jkeenan/testing/blead/lib/perl5/5.35.9/amd64-freebsd-thread-multi/CORE"  -DNDEBUG -DHAVE_ZSTD Decoder.c
In file included from /home/jkeenan/testing/blead/lib/perl5/5.35.9/amd64-freebsd-thread-multi/CORE/perl.h:7162,
                 from Decoder.xs:6:
/home/jkeenan/testing/blead/lib/perl5/5.35.9/amd64-freebsd-thread-multi/CORE/inline.h: In function 'Perl_cop_file_avn':
/home/jkeenan/testing/blead/lib/perl5/5.35.9/amd64-freebsd-thread-multi/CORE/inline.h:3439:5: error: ISO C90 forbids mixed declarations and code [-Werror=declaration-after-statement]
 3439 |     const char *file = CopFILE(cop);
      |     ^~~~~
cc1: some warnings being treated as errors
*** Error code 1

Stop.
make: stopped in /usr/home/jkeenan/.cpan/build/Sereal-Decoder-4.018-3
  YVES/Sereal-Decoder-4.018.tar.gz
  /usr/bin/make -- NOT OK
```
Same box, same commit; only thing that's different is the C-compiler used.

On an older Linux box where `gcc-4.9.2` is the default C-compiler, I was able to reproduce the `Sereal::Decoder` build failures reported by @andk.  When I used `clang` as the C-compiler, I got:
```
$ ./bin/perl -v | head -2 | tail -1
This is perl 5, version 35, subversion 9 (v5.35.9 (v5.35.8-29-g4ccd151)) built for x86_64-linux-thread-multi
$ ./bin/perl -V:config_args
config_args='-des -Dusedevel -Duseithreads -Dcc=clang -Dprefix=/home/jkeenan/testing/blead -Uversiononly -Dman1dir=none -Dman3dir=none';
$ clang --version
Debian clang version 3.5.0-10 (tags/RELEASE_350/final) (based on LLVM 3.5.0)
Target: x86_64-pc-linux-gnu
Thread model: posix
$ ./bin/cpan -i Sereal::Decoder
...
$ ./bin/perl -Ilib -MSereal::Decoder -E 'say qq|$Sereal::Decoder::VERSION|;'
4.018
```

**Comment by jkeenan** [comment]
This Perl compilation change, documented in `pod/perl5355delta.pod`, may be relevant:
```
Configuration and Compilation
*   The Perl C source code now uses some C99 features, which we have
    verified are supported by all compilers we target. This means that
    Perl's headers now contain some code that is legal in C99 but not C89.

    This may cause problems for some XS modules that unconditionally add
    "-Werror=declaration-after-statement" to their C compiler flags if
    compiling with gcc or clang. Earlier versions of Perl support long
    obsolete compilers that are strict in rejecting certain C99 features,
    particularly mixed declarations and code, and hence it makes sense for
    XS module authors to audit that their code does not violate this.
    However, doing this is now only possible on these earlier versions of
    Perl, hence these modules need to be changed to only add this flag for
    "<$] < 5.035005">.
 ```

Relevant part of the breaking commit:
```
diff --git a/inline.h b/inline.h
index e6fd4cb6a9..41da923991 100644
--- a/inline.h
+++ b/inline.h
@@ -3429,6 +3429,28 @@ Perl_sv_isbool(pTHX_ const SV *sv)
         (SvPVX_const(sv) == PL_Yes || SvPVX_const(sv) == PL_No);
 }
 
+#ifdef USE_ITHREADS
+
+PERL_STATIC_INLINE AV *
+Perl_cop_file_avn(pTHX_ const COP *cop) {
+
+    PERL_ARGS_ASSERT_COP_FILE_AVN;
+
+    const char *file = CopFILE(cop);
+    if (file) {
+        GV *gv = gv_fetchfile_flags(file, strlen(file), GVF_NOADD);
+        if (gv) {
+            return GvAVn(gv);
+        }
+        else
+            return NULL;
+     }
+     else
+         return NULL;
+}
+
+#endif
+
```

@tonycoz, @Leont, @demerphq: can you take a look?

Thank you very much.
Jim Keenan

**Comment by Leont** [comment]
Probably the dist should just stop adding `-Werror=declaration-after-statement` [here](https://metacpan.org/release/YVES/Sereal-Decoder-4.018/source/inc/Sereal/BuildTools.pm#L186)

**Comment by jkeenan** [comment]
> Probably the dist should just stop adding `-Werror=declaration-after-statement` [here](https://metacpan.org/release/YVES/Sereal-Decoder-4.018/source/inc/Sereal/BuildTools.pm#L186)

I guess the larger question is:   What part of the fix needs to be done in core?  What part in the CPAN distro?

For an attempt at a fix in core, see https://github.com/Perl/perl5/pull/19384.

**Comment by jkeenan** [comment]
@demerphq released Sereal-Decoder-4.023 to CPAN on Feb 19 2022, which appears to have cleared up the problem.
```
CPAN::Reporter: Test result is 'pass', 'make test' no errors.
CPAN::Reporter: preparing a CPAN Testers report for Sereal-Decoder-4.023

CPAN::Reporter: this appears to be a duplicate report for the test phase:
PASS Sereal-Decoder-4.023 amd64-freebsd-thread-multi 13.0-release
```
Closing ticket.
